src_c/transform.c

```diff
@@ -2118,6 +2118,225 @@ surf_grayscale(PyObject *self, PyObject *args, PyObject *kwargs)
+    Uint8 r, g, b, a;
+    Uint8 *src_pixels = (Uint8 *)surf->pixels, *pix;
+    Uint8 *dst_pixels = (Uint8 *)newsurf->pixels;
+    float s_h = 0, s_s = 0, s_l = 0;
+    Uint32 pixel;
+
+    for (y = 0; y < surf->h; y++) {
+        for (x = 0; x < surf->w; x++) {
+            SURF_GET_AT(pixel, surf, x, y, src_pixels, surf->format, pix);
+            SDL_GetRGBA(pixel, surf->format, &r, &g, &b, &a);
+            RGB_to_HSL(r, g, b, &s_h, &s_s, &s_l);
+
+            if (h) {
+                s_h += h;
+                if (s_h > 1)
+                    s_h -= 1;
+                else if (s_h < 0)
+                    s_h += 1;
+            }
+            if (s) {
+                s_s = s_s * (1 + s);
+                s_s = s_s > 1 ? 1 : s_s < 0 ? 0 : s_s;
+            }
+            if (l) {
+                if (l < 0) {
+                    s_l = s_l * (1 + l);
+                }
+                else {
+                    s_l = s_l * (1 - l) + l;
+                }
+                s_l = s_l > 1 ? 1 : s_l < 0 ? 0 : s_l;
+            }
+
+            HSL_to_RGB(s_h, s_s, s_l, &r, &g, &b);
+            pixel = SDL_MapRGBA(newsurf->format, r, g, b, a);
+            SURF_SET_AT(pixel, newsurf, x, y, dst_pixels, newsurf->format,
+                        pix);
+        }
+    }
+
+    SDL_UnlockSurface(newsurf);
+    return newsurf;
+}
+
+static PyObject *
+surf_hsl(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    SDL_Surface *surf = pgSurface_AsSurface(self);
+    SURF_INIT_CHECK(surf)
+    pgSurfaceObject *surfobj;
+    pgSurfaceObject *surfobj2 = NULL;
+    SDL_Surface *newsurf;
+    float h = 0, s = 0, l = 0;
+
+    static char *keywords[] = {"surface",   "hue",          "saturation",
+                               "lightness", "dest_surface", NULL};
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O!|fffO!", keywords,
+                                     &pgSurface_Type, &surfobj, &h, &s, &l,
+                                     &pgSurface_Type, &surfobj2)) {
+        return NULL;
+    }
+
+    if (s < -1 || s > 1) {
+        PyObject *value = PyFloat_FromDouble((double)s);
+        if (!value) {
+            return NULL;
+        }
+        PyErr_Format(PyExc_ValueError,
+                     "saturation value must be between -1 and 1, got %R",
+                     value);
+        Py_XDECREF(value);
+        return NULL;
+    }
+    if (l < -1 || l > 1) {
+        PyObject *value = PyFloat_FromDouble((double)l);
+        if (!value) {
+            return NULL;
+        }
+        PyErr_Format(PyExc_ValueError,
+                     "lightness value must be between -1 and 1, got %R",
+                     value);
+        Py_XDECREF(value);
+        return NULL;
+    }
+
+    if (h > 0) {
+        while (h > 360) {
+            h -= 360;
+        }
+    }
+    else if (h < 0) {
+        while (h < -360) {
+            h += 360;
+        }
+    }
+    h /= 360.0;
+
+    newsurf = modify_hsl(surfobj, surfobj2, h, s, l);
```

Comment:
```suggestion
    pgSurface_Lock(surfobj);
    Py_BEGIN_ALLOW_THREADS;
    newsurf = modify_hsl(surfobj, surfobj2, h, s, l);
    Py_END_ALLOW_THREADS;
    pgSurface_Unlock(surfobj);
```

Do you mean something like this?

Reply:
This was meant in reply to @yunline

Reply:
> Do you mean something like this?

Yes that is what I mean.

Reply:
I added the thread stuff but seems like it led to errors. I don't know wy tbh.